channels/rdpecam/client/v4l/camera_v4l.c

```diff
@@ -0,0 +1,727 @@
+{
+	switch (ecamFormat)
+	{
+		case CAM_MEDIA_FORMAT_H264:
+			return V4L2_PIX_FMT_H264;
+		case CAM_MEDIA_FORMAT_MJPG:
+			return V4L2_PIX_FMT_MJPEG;
+		case CAM_MEDIA_FORMAT_YUY2:
+			return V4L2_PIX_FMT_YUYV;
+		case CAM_MEDIA_FORMAT_NV12:
+			return V4L2_PIX_FMT_NV12;
+		case CAM_MEDIA_FORMAT_I420:
+			return V4L2_PIX_FMT_YUV420;
+		case CAM_MEDIA_FORMAT_RGB24:
+			return V4L2_PIX_FMT_RGB24;
+		case CAM_MEDIA_FORMAT_RGB32:
+			return V4L2_PIX_FMT_RGB32;
+		default:
+			WLog_ERR(TAG, "Unsupported CAM_MEDIA_FORMAT %d", ecamFormat);
+			return 0;
+	}
+}
+
+/**
+ * Function description
+ *
+ * @return file descriptor
+ */
+static int cam_v4l_open_device(const char* deviceId, int flags)
+{
+	UINT n;
+	char device[20];
+	int fd = -1;
+	struct v4l2_capability cap;
+
+	if (!deviceId)
+		return -1;
+
+	if (0 == strncmp(deviceId, "/dev/video", 10))
+		return open(deviceId, flags);
+
+	for (n = 0; n < 64; n++)
+	{
+		snprintf(device, sizeof(device), "/dev/video%d", n);
+		if ((fd = open(device, flags)) == -1)
+			continue;
+
+		/* query device capabilities and make sure this is a video capture device */
+		if (ioctl(fd, VIDIOC_QUERYCAP, &cap) < 0 ||
+			!(cap.device_caps & V4L2_CAP_VIDEO_CAPTURE))
+		{
+			close(fd);
+			continue;
+		}
+
+		if (cap.bus_info[0] != 0 && 0 == strcmp((const char*)cap.bus_info, deviceId))
+			return fd;
+
+		close(fd);
+	}
+
+	return fd;
+}
+
+/**
+ * Function description
+ *
+ * @return -1 if error, otherwise index of supportedFormats array and mediaTypes/nMediaTypes filled in
+ */
+static INT16 cam_v4l_get_media_type_descriptions(ICamHal* hal, const char* deviceId, int streamIndex,
+		const CAM_MEDIA_FORMAT_INFO* supportedFormats, size_t nSupportedFormats,
+		CAM_MEDIA_TYPE_DESCRIPTION* mediaTypes, size_t* nMediaTypes)
+{
+	int fd;
+	size_t maxMediaTypes = *nMediaTypes;
+	size_t nTypes = 0; 
+	int formatIndex;
+	BOOL formatFound = FALSE;
+	struct v4l2_format video_fmt = { 0 };
+
+	video_fmt.type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
+	video_fmt.fmt.pix.sizeimage = 0;
+
+	unsigned int videoSizes[][2] =
+		{{160, 90}, {160, 120}, {320, 180}, {320, 240}, {432, 240}, {352, 288}, {640, 360},
+		{800, 448}, {640, 480}, {848, 480}, {864, 480}, {960, 540}, {1024, 576}, {800, 600},
+		{960, 720}, {1280, 720}, {1024, 768}, {1600, 896}, {1440, 1080}, {1920, 1080}};
+	const int totalSizes = sizeof(videoSizes) / sizeof(unsigned int[2]);
+
+	if ((fd = cam_v4l_open_device(deviceId, O_RDONLY)) == -1)
+	{
+		WLog_ERR(TAG, "Unable to open device %s", deviceId);
+		return -1;
+	}
+
+	for (formatIndex = 0; formatIndex < nSupportedFormats; formatIndex++)
+	{
+		UINT32 pixelFormat = ecamToV4L2PixFormat(supportedFormats[formatIndex].inputFormat);
+		WINPR_ASSERT(pixelFormat != 0);
+		
+		for (int i = 0; i < totalSizes; i++)
+		{
+			video_fmt.fmt.pix.pixelformat = pixelFormat;
+			video_fmt.fmt.pix.width = videoSizes[i][0];
+			video_fmt.fmt.pix.height = videoSizes[i][1];
+
+			if (ioctl(fd, VIDIOC_TRY_FMT, &video_fmt) >= 0)
```

Comment:
May I suggest 
```C
if (ioctl(fd, VIDIOC_TRY_FMT, &video_fmt) < 0 || 
        video_fmt.fmt.pix.pixelformat != pixelFormat ||
	video_fmt.fmt.pix.width != videoSizes[i][0] ||
	video_fmt.fmt.pix.height != videoSizes[i][1])
   continue;
```

to avoid the staircase ?

Reply:
Done